extras/lbf-nix/lbf-typescript.nix

```diff
@@ -0,0 +1,282 @@
+  files
+  # Classes for which to generate implementations for (default lbf-prelude classes).
+, classes ? [ ]
+, # npmExtraDependencies are tarballs from `npm pack` of the dependencies required
+  npmExtraDependencies ? [ ]
+  # , # npmDevDependencies are tarballs from `npm pack` of the develoepr dependencies for building the project
+  #   # TODO(jaredponn): actually use this
+  #   npmDevDependencies ? [ ]
+, configs ? [ ]
+, # Name of the package and also the name of the Cabal package.
+  # Examples: name = "lbf-myproject"
+  name
+, # Version of the package and also the version of the Cabal package.
+  # Examples: version = "0.1.0.0"
+  version ? "1.0.0"
+
+}:
+let
+  lbf-build = import ./lbf-build.nix pkgs lbf;
+
+  # TODO(jaredponn): this was stolen from the Rust side. This should be made a
+  # common function.
+  findModules = root: builtins.map
+    (path: builtins.replaceStrings [ "/" ] [ "." ]
+      (pkgs.lib.strings.removePrefix "./" (pkgs.lib.strings.removeSuffix ".lbf"
+        (pkgs.lib.path.removePrefix root path))))
+    (builtins.filter (pkgs.lib.hasSuffix ".lbf")
+      (pkgs.lib.filesystem.listFilesRecursive root));
+
+  packageSet =
+    pkgs.writeTextFile {
+      name = "lb-typescript-packages";
+      text =
+        builtins.toJSON
+          (builtins.mapAttrs (_name: findModules) imports);
+    };
+
+  lbfBuilt = with lbfTypescriptOpts;
+    lbf-build.build
+      {
+        inherit src;
+        opts = {
+          inherit files;
+          import-paths = pkgs.lib.attrsets.attrValues imports;
+          gen = lbg-typescript;
+          gen-classes = classes;
+          gen-dir = "autogen";
+          gen-opts =
+            [ "--packages=${packageSet}" ]
+            ++ builtins.map (c: "--config=${c}") configs;
+          # WARN(bladyjoker): If I put quotes here everything breaks.
+          work-dir = ".work";
+        };
+      };
+
+  packageJsonTemplate = with lbfTypescriptOpts;
+    pkgs.writeTextFile {
+      name = "lambda-buffers-package-json-template";
+      text =
+        ''
+          {
+            "name": "${name}",
+            "version": "${version}",
+            "description": "A package.json that contains LambdaBuffers generated TypeScript files",
+            "lockfileVersion": 2,
+            "exports": {
+              "./package.json": "./package.json"
+            },
+            "type": "module",
+            "scripts": {
+              "build": "npx tsc -b .",
+              "test": ":"
+            },
+            "devDependencies": {
+            },
+            "files": ["./dist/LambdaBuffers/**/*" ],
+            "dependencies": {
+            }
+          }
+        '';
+    };
+
+  # TODO(jaredponn): urgh, this was manually copy pastad in by hand.. automate
+  # this more nicely later... there's a huge pain point with how this needs to
+  # hit the network to create the package-lock.json...
+  packageJsonLockTemplate = with lbfTypescriptOpts;
```

Comment:
I see this is sad :D Updating the TS stack then requires changing this?

Reply:
Yes, you understand the (rather bleak!) situation perfectly. Happy to hear some suggestions to avoid this! 

Frankly, this is just part of the pain points I've been having with `npm` in general -- it also does the same hashing / package management thing that nix does; so these systems are a bit "at odds" with each other....